src/deepmol/pipeline/pipeline.py

```diff
@@ -121,13 +131,33 @@ def _fit(self, dataset: Dataset) -> 'Pipeline':
         """
         self._validate_steps()
 
-        if self.is_prediction_pipeline():
-            for name, transformer in self.steps[:-1]:
-                dataset = transformer.fit_transform(dataset)
-            self.steps[-1][1].fit(dataset)
+        if self.hpo is None:
+            if self.is_prediction_pipeline():
+                for name, transformer in self.steps[:-1]:
+                    train_dataset = transformer.fit_transform(train_dataset)
+                self.steps[-1][1].fit(train_dataset)
+            else:
+                for name, transformer in self.steps:
+                    train_dataset = transformer.fit_transform(train_dataset)
         else:
-            for name, transformer in self.steps:
-                dataset = transformer.fit_transform(dataset)
+            if self.is_prediction_pipeline():
+                for name, transformer in self.steps[:-1]:
+                    train_dataset = transformer.fit_transform(train_dataset)
+                    if validation_dataset is not None:
+                        validation_dataset = transformer.transform(validation_dataset)
+            else:
+                for name, transformer in self.steps:
+                    train_dataset = transformer.fit_transform(train_dataset)
+                    if validation_dataset is not None:
+                        validation_dataset = transformer.transform(validation_dataset)
+
+            best_model, self.hpo_best_hyperparams_, self.hpo_all_results_ = self.hpo.fit(train_dataset,
+                                                                                         validation_dataset)
+            if self.is_prediction_pipeline():
```

Comment:
we need to assure that refit=True

Reply:
I checked the code and it seems that if the user sets refit=False, it maintains the best_estimator of the CV, whereas if it is True, it retrains the model with the whole dataset... I think it is the expected behaviour, no?

Reply:
Great. That's it.